Importing database from SQL

JorisGoosen opened the pull request:
Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1693
Fixes https://github.com/jasp-stats/jasp-issues/issues/29

JorisGoosen committed:
Started work on importing datasets from a database

hide password #and some other stuff
add port to settings

importing JASP is nice

JorisGoosen committed:
Can actually connect to a database

JorisGoosen committed:
Importing Database, the babysteps

A whole bunch of untested code for sharing information about a database connection within JASP.
Also setting up the import-path for it

JorisGoosen committed:
Fleshing out the importer

it sadly enough does not want to connect to my db...

JorisGoosen committed:
Importing from DB works! :D

JorisGoosen committed:
can synchronize now

JorisGoosen committed:
GUI straightened out a bit and can save+load to jasp-file now

JorisGoosen committed:
Adds an interval for synching with the database

JorisGoosen committed:
Fix typo to import double

JorisGoosen committed:
Fix dbtype combobox

JorisGoosen committed:
scrolling in database menu and better cmake

JorisGoosen committed:
only add arguments to QDB if filled

JorisGoosen committed:
dont show warning again and ErrorMessage Widget

JorisGoosen committed:
add some documentation - sqlite example

JorisGoosen committed:
apparently sqlite first returns an empty record

JorisGoosen committed:
Actually add the md file

JorisGoosen commented:
@boutinb suggested adding a "remember me" checkmark like at OSF which could work like:
-   check remember me -> stores password in settings and in JASP-file if saved.
 -  dont check remember me -> uses password + user while loaded. but it isnt saved and everytime the file is loaded the user is prompted with a user+password dialog?

boutinb commented on Desktop/components/JASP/Widgets/ErrorMessage.qml:
```diff
@@ -0,0 +1,100 @@
+		}
+	}
+
+	CrossButton { onCrossClicked: {	errorMsg.opacity = 0 } }
```
By clicking the cross button, this will just make the message disappear, but it is not removed: the place it takes will be still used, making a big empty place.

JorisGoosen replied:
oops

boutinb commented on Desktop/components/JASP/Widgets/FileMenu/Database.qml:
```diff
@@ -0,0 +1,379 @@
+                }
+            }
+
+            RoundedButton
```
This button should appear only when the database is not yet connected. When connected it should show 'Close database', and the connection settings should be disabled.

JorisGoosen replied:
Ok but I think this is a fundamental thing here, the "Open" menu is not for changing the currently loaded dataset at all.
So a database is never "connected" once you enter that screen right?

What would the behaviour even be?
You press "Load into JASP" and then you lose all changes you made? 
Why would you change the connection once it works?

boutinb commented on Desktop/components/JASP/Widgets/FileMenu/Database.qml:
```diff
@@ -0,0 +1,379 @@
+                }
+
+                SpinBox
+                {
```
If I change back the synchronization to 0, then it still tries to synchronize.

JorisGoosen replied:
so you mean after you loaded it? Again the same open != change-current

boutinb commented on Desktop/components/JASP/Widgets/FileMenu/Database.qml:
```diff
@@ -0,0 +1,379 @@
+                    value:				fileMenuModel.database.interval
+                    onValueChanged:		if(value != "") fileMenuModel.database.interval = value
+                    from:				0
+                    to:					24 * 60 * 7 // do we want to allow for an interval bigger than a week?
```
Why to put a max limitation? It's not necessary, you never know how JASP will be used in te future.

JorisGoosen replied:
I guess, I will remove it.

boutinb commented on Desktop/components/JASP/Widgets/FileMenu/Database.qml:
```diff
@@ -0,0 +1,379 @@
+                }
+
+                RoundedButton
+                {
```
If I change the query, execute it, and load it again, then it does not change the data in JASP.

JorisGoosen replied:
and it also doesnt open a new JASP?
because that is what I would expect

boutinb commented on Desktop/components/JASP/Widgets/FileMenu/Database.qml:
```diff
@@ -0,0 +1,379 @@
+        {
+            id:		databaseGroup
+
+            title:	qsTr("Database")
```
Are all fields needed for all kinds of database? I don't think a port is needed for SQLLite for example.

JorisGoosen replied:
In fact you could leave empty what you want and port==0 is empty there.
sqlite for sure doesnt need that indeed, I also mention this in the guide: https://github.com/JorisGoosen/jasp-desktop/blob/importDB/Resources/Help/filemenu/Database.md#sqlite

boutinb reviewed (changes requested):
It works really easily, but the interface can be fine-tuned. Especially when having loaded the data, and you come back to the Open/Database, then it is not clear what you can change or not (in fact any change does not seem to work..).

boutinb commented:
Also, if I modify the SQL statement, save it as JASP file, and open the JASP file again, it still uses the old settings.
Another point the Sync Data menu, should be disabled when a Database connection is used, or the Synchronization setting should be set in this menu.

JorisGoosen commented:
Well, using the open-interface would result in a new dataset being loaded right?
There is no provision made for changing the database connection in JASP later... 

So changing the query would also have no effect on the currently loaded JASP. But if you press "Load into JASP" it *should* open a new dataset (but I dont think Ive tested that).

JorisGoosen committed:
add db to sync and add a nonfunctional "remember me" checkbox

JorisGoosen committed:
halfbaked commit preparing for actually remembering and reasking the password if not

JorisGoosen committed:
Fix some postrebase errors etc

JorisGoosen committed:
add documentation on how to connect with sql server

JorisGoosen committed:
Remember me sort of works now

JorisGoosen committed:
still something going wrong, but improvements are made

JorisGoosen committed:
Only synch if there is an interval

JorisGoosen committed:
fix savebutton being enabled randomly and wrongly
reset database qml fields after loading into jasp
dont close the filemenu after closing a jasp-workspace

JorisGoosen committed:
make sure we can start a separate instance of jasp that then loads from a databaseconnection

JorisGoosen committed:
instead of having several locations to store the current "FileMode" just store it in FileMenu only and access it through a pointer

JorisGoosen committed:
Adds support for synchronize interface

And general fixes here and there

JorisGoosen committed:
no maximum interval please

JorisGoosen commented:
Well, @boutinb is on vacation and I think all the prvevious issues have been resolved so Im just gonna merge it.
